python/nav/bin/navclean.py

```diff
@@ -195,18 +220,54 @@ def delete(self, expiry, dry_run=False):
 # pylint: disable=missing-docstring
 
 
-class ArpDeleter(RecordDeleter):
-    table = "arp"
+class ArpDeleter(RecordCleaner):
+    expiry_type = "arp"
     selector = "WHERE end_time < {expiry}"
 
 
-class CamDeleter(RecordDeleter):
-    table = "cam"
+class ArpCloser(RecordCleaner):
+    """Closes ARP records that have "expired", i.e. they're open, but the netbox they
+    were collected from has been down for too long.
+    """
+
+    expiry_type = "close_arp"
+
+    def sql(self, expiry):
+        """Returns the full UPDATE statement based on the expiry date"""
+        return f"""
+            WITH downsince AS (
+                SELECT
+                    netboxid,
+                    start_time AS downsince
+                FROM
+                    alerthist
+                WHERE
+                    eventtypeid = 'boxState'
+                    AND end_time >= 'infinity'
+            )
+            UPDATE
+                arp
+            SET
+                end_time = NOW()
+            WHERE
+                netboxid IN (
+                    SELECT
+                        netboxid
+                    FROM
+                        downsince
+                    WHERE
+                        downsince < {expiry})
+                AND end_time >= 'infinity'
+        """
```

Comment:
Been a while since ive done any serious sql-ing myself, but this does seem to be fine. Could consider doing the `downsince < {expiry})` comparison in the first part so less data is used for comparison later, but it probably doesnt matter much.

the double `downsince` naming was also a little bit confusing

Reply:
Agree about the confusing double-downsince.  Other than that, the PostgreSQL query planner should normally be smart enough to optimize this query based on the subsequent join, so it's written like this to be easier to read for the next dev that touches it.

Reply:
I know very little about the psql query planner so I'll choose to believe you :stuck_out_tongue: 

New names are much better :+1: